### PR TITLE
dominating_set function randomization

### PR DESCRIPTION
**Pull request by goznalo-git**
Hi there! 
First attempt to contribute to the project, so please let me know anything I might have missed or should do differently. 

Regarding the PR: I have modified the `networkx/algorithms/dominating.py` file, introducing the possibility of a randomization in the procedure of obtaining the dominating set. This allows finding different sets every time the function is called with the parameter set to True. I have modified the function description accordingly.
- I ran my own tests obtaining correct sets (checked agains the `is_dominating_set()` function) of randomly generated digraphs.
- I also ran the black pre-commit hook, and the pytest tests.

Thanks for your attention! Looking forward to your feedback.
Gonzalo

### CODE REVIEW

**Comment by dschult** [comment]
Sorry this took so long to get to...

This is well formatted and applied using git. It should also have some tests and the doc_strings should restrict their line lengths to about 80 chars. 

But I have trouble with the algorithm -- 
If you want to explore many different outcomes, perhaps it is better to construct a method that explores the different results systematically rather than randomly. 
On the other hand, if you are introducing a random aspect to the result, shouldn't you be able to tell what the distribution of the possible outcomes are? Maybe that is too much to ask, but you should at least think about it. The implemented randomness may sample the possible outcomes in a very non-uniform manner.  Do you have a feel for what distribution the outcomes have?

Finally, your desired outcome might be better achieved by shuffling the node order of the graph. By that, I mean you could reorder the nodes so that they are stored in a different order on the data structure and run the resultant graph through the deterministic algorithm.  Something like:
```python
nodes, edges = list(G.nodes), list(G.edges)
random.shuffle(nodes)
random.shuffle(edges)

newG = G.__class__()
newG.add_nodes_from(nodes)
newG.add_edges_from(edges)
```

Thoughts?

**Comment by goznalo-git** [comment]
Hi @dschult, 

Thank you for your feedback! Before making any changes to the PR, I think we should discuss your comments.

In particular, I think it will be worth telling you my motivation behind this, which is twofold:
- On the one hand, I wanted to contribute to Networkx as I'll likely be using it during my PhD, and as such this was a first attempt to help, and understand what is expected in PRs and things like that. 
- On the other hand I chose this particular change in preparation for another, bigger, addition I have in mind. In [https://arxiv.org/abs/1109.4521](https://arxiv.org/abs/1109.4521), the authors comment that finding the minimum controlling (dominating) set is an NP-hard problem, thus they can only rely on two algorithms (TDCS, BDCS), arguing that they have found them to output smaller sets than randomized counterparts.
I thought it might be interesting having these algorithms implemented in Networkx, as well as the random version. I started with the latter due to its simplicity, but my plan is to go all the way, coding the Top-Down and Bottom-Up ones.

In summary, I am not that interested in the properties, behaviour, etc of the randomized version, as it is somewhat collateral, and would like to keep it as simple as possible. I am also not sure about your inquiries, or how they'd come into play here. For instance, a systematic search is opposed to what I strive for, which is a non-deterministic outcome. TDCS, BDCS will actually be deterministic.

Please let me know what you think about this. 

Best, 
G.

**Comment by rossbar** [comment]
Thanks for sharing the additional context @goznalo-git 

> If you want to explore many different outcomes, perhaps it is better to construct a method that explores the different results systematically rather than randomly.

I had the same initial reaction as well. Another consideration re: "randomizing" the output is that, if we were to add this feature, we would also need a way to make the randomized output reproducible. Typically this is done by providing a `seed` kwarg and using the random state decorator.

To me though, the randomization feels like it would be best handled outside the function, which can be done straightforwardly w/ shuffling as suggested by @dschult . This is a clean solution that AFAICT would give the desired behavior and has the additional advantage of not requiring an expansion to the API.

**Comment by goznalo-git** [comment]
Hi @rossbar,

Thank you for your comments. I think I now agree with both of you, the random shuffling option might be the better choice, indeed. Still, before closing this PR, do you and @dschult  think I should go on and code the TDCS and BDCS algorithms, as I intended? Those are AFAIK not as easy to do outside the API as the random version, and they are deterministic. If so I'll work on them on a separate PR

Thanks!

**Comment by rossbar** [comment]
> do you and @dschult think I should go on and code the TDCS and BDCS algorithms, as I intended? 

I'm not familiar with these algorithms so I can't say :). From an organizational perspective though, I'd recommend opening a new [algorithm discussion topic](https://github.com/networkx/networkx/discussions/categories/algorithms) so that the discussion is framed with the appropriate context and more visible to those who might miss this thread in the PR. You could either copy/paste the relevant parts of [your comment above](https://github.com/networkx/networkx/pull/4803#issuecomment-860252245) or simply link to it from the new discussion if that's easier!

**Comment by dschult** [comment]
I think opening an algorithm discussion might be in order.  In regard to your specific question about implementing top-down controller set (TDCS) and bottom-up.  As far as I can tell, these methods are only mentioned in the last few paragraphs of that paper from 2011 -- not the heart of the study.  And with no further description or links that I can find, it is hard to justify including them.  That said, it might help **you** in your studies to go ahead and implement these algorithms and play with them to see if you get useful results. Who knows, you might find out something! :}  But -- I would suggest that you read more broadly before choosing which algorithms to implement. Perhaps these **are** of interest and I just wasn't successful in finding that literature. Or perhaps there is a more interesting algorithm for you to implement.
